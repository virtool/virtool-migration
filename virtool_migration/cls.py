from dataclasses import dataclass
from datetime import datetime
from typing import Callable


@dataclass
class AppliedRevision:
    id: str
    created_at: datetime
    name: str


@dataclass
class Revision(AppliedRevision):
    upgrade: Callable
