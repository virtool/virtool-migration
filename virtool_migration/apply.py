from logging import getLogger
from pathlib import Path
from typing import List

import arrow
from alembic.util import load_python_file
from motor.motor_asyncio import AsyncIOMotorClient

from virtool_migration.cls import Revision

logger = getLogger(__name__)


async def apply_to(mongo_connection_string: str, revisions_path: Path, to: str):
    """
    Apply revisions up to a given revision_id, ``to``.

    Providing ``'latest'`` as for ``to`` will apply all required revisions up to latest.

    Applied revisions are recorded in the target database so already applied revisions
    are not reapplied in subsequent migrations.

    :param mongo_connection_string: the MongoDB connection string
    :param revisions_path: the path to the directory containing revision modules
    :param to: the revision_id to update to or 'latest'
    """
    logger.info("Applying revisions up to '%s'", to)

    revisions = _load_revisions(revisions_path)

    motor_client = AsyncIOMotorClient(mongo_connection_string)

    for revision in revisions:
        await apply_revision(motor_client, revision)

        if revision != "latest" and revision.id == to:
            break


async def apply_revision(motor_client: AsyncIOMotorClient, revision: Revision):
    logger.info("Applying revision '%s' (%s)", revision.name, revision.id)

    async with await motor_client.start_session() as session:
        async with session.start_transaction():
            await revision.upgrade(motor_client.get_default_database(), session)
            await motor_client.get_default_database().migrations.insert_one(
                {
                    "applied_at": arrow.utcnow().naive,
                    "created_at": revision.created_at,
                    "name": revision.name,
                    "revision_id": revision.id,
                },
                session=session,
            )


def _load_revisions(revisions_path: Path) -> List[Revision]:
    """
    Load revision modules from the ``revisions_path`` and sort by creation date.

    :param revisions_path: the path to the revisions directory
    :returns: a list of revisions sorted by creation date

    """
    revisions = []

    for module_path in revisions_path.iterdir():
        if module_path.suffix == ".py":
            module = load_python_file(str(module_path.parent), str(module_path.name))

            revisions.append(
                Revision(
                    id=getattr(module, "revision_id"),
                    created_at=arrow.get(getattr(module, "created_at")).datetime,
                    name=getattr(module, "name"),
                    upgrade=getattr(module, "upgrade"),
                )
            )

    revisions.sort(key=lambda r: r.created_at)

    logger.info("Found %s revisions", len(revisions))

    return revisions
