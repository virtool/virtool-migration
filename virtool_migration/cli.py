import asyncio
import logging
from pathlib import Path
from subprocess import call
from typing import Optional

import click

from virtool_migration.apply import apply_to
from virtool_migration.revisions import create_revision

logging.basicConfig(
    level=logging.DEBUG,
)


def entry():
    cli(auto_envvar_prefix="VT_MIGRATION", obj={})


@click.group()
@click.pass_context
def cli(ctx):
    ctx.ensure_object(dict)


@cli.command()
@click.option(
    "--path",
    help="The path to the revisions directory",
    default=Path.cwd() / "revisions",
    type=Path,
)
@click.option(
    "--mongo-connection-string",
    help="The connection string for the MongoDB database to migrate",
    default="mongodb://localhost:27017/virtool",
)
@click.option(
    "--revision-id",
    help="Apply migrations up to and including this revision id",
    default="latest",
)
@click.option("--alembic", help="Run Alembic migrations up to head", default=False)
def apply(mongo_connection_string: str, path: Path, revision_id: str, alembic: bool):
    asyncio.run(apply_to(mongo_connection_string, path, revision_id))

    if alembic:
        call(["alembic", "upgrade", "header"])


@cli.command()
@click.option("-n", "--name", help="A human-readable name for the revision")
@click.option(
    "--path",
    help="The path to the revisions directory",
    default=Path.cwd() / "revisions",
    type=Path,
)
def new(name: Optional[str], path: Path):
    create_revision(path, name)
