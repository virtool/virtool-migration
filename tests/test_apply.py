from datetime import datetime

import pytest
from syrupy.filters import props
from syrupy.matchers import path_type

from virtool_migration.apply import apply_to


@pytest.fixture
async def examples(mongo):
    await mongo.examples.insert_many(
        [
            {
                "groups": [
                    "fred",
                    "techs",
                ],
                "username": "fred",
            },
            {
                "groups": [
                    "bob",
                    "techs",
                ],
                "username": "bob",
            },
            {
                "groups": [
                    "lisa",
                ],
                "username": "lisa",
            },
        ]
    )


async def test_apply_to_latest(
    examples, full_mongo_connection_string, mongo, revisions_path, snapshot
):

    await apply_to(full_mongo_connection_string, revisions_path, to="latest")

    assert await mongo.examples.find().to_list(None) == snapshot(
        exclude=props(
            "_id",
        )
    )

    assert await mongo.migrations.find().to_list(None) == snapshot(
        exclude=props("_id"),
        matcher=path_type(
            {
                ".*applied_at": (datetime,),
            },
            regex=True,
        ),
    )


async def test_apply_in_two_stages(
    examples, full_mongo_connection_string, mongo, revisions_path, snapshot
):
    """
    Test that:

    1. Using `to` to point to a specific revision only results in the desired changes
       being applied.
    2. Following up with a `to=latest` results in only the remaining migration being
       applied.

    """
    await apply_to(full_mongo_connection_string, revisions_path, to="mujt2zxouf9p")

    assert await mongo.examples.find().to_list(None) == snapshot(
        exclude=props(
            "_id",
        ),
        name="first_examples",
    )

    migrations = await mongo.migrations.find().to_list(None)

    assert len(migrations) == 1

    first_migration = migrations[0]

    assert migrations == snapshot(
        exclude=props("_id"),
        matcher=path_type(
            {
                ".*applied_at": (datetime,),
            },
            regex=True,
        ),
        name="first_migrations",
    )

    await apply_to(full_mongo_connection_string, revisions_path, to="latest")

    assert await mongo.examples.find().to_list(None) == snapshot(
        exclude=props(
            "_id",
        ),
        name="second_examples",
    )

    migrations = await mongo.migrations.find().to_list(None)

    assert migrations == snapshot(
        exclude=props("_id"),
        matcher=path_type(
            {
                ".*applied_at": (datetime,),
            },
            regex=True,
        ),
        name="second_migrations",
    )

    assert migrations[0] == first_migration
