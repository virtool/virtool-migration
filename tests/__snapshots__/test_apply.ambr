# name: test_apply_in_two_stages[first_examples]
  list([
    dict({
      'groups': list([
        'fred',
        'techs',
      ]),
      'user_id': 'fred',
    }),
    dict({
      'groups': list([
        'bob',
        'techs',
      ]),
      'user_id': 'bob',
    }),
    dict({
      'groups': list([
        'lisa',
      ]),
      'user_id': 'lisa',
    }),
  ])
# ---
# name: test_apply_in_two_stages[first_migrations]
  list([
    dict({
      'applied_at': datetime,
      'created_at': datetime.datetime(2022, 6, 8, 16, 49, 51),
      'name': 'Rename username to user_id',
      'revision_id': 'mujt2zxouf9p',
    }),
  ])
# ---
# name: test_apply_in_two_stages[second_examples]
  list([
    dict({
      'archived': False,
      'groups': list([
        'fred',
        'techs',
      ]),
      'user': 'fred',
    }),
    dict({
      'archived': False,
      'groups': list([
        'bob',
        'techs',
      ]),
      'user': 'bob',
    }),
    dict({
      'archived': False,
      'groups': list([
        'lisa',
      ]),
      'user': 'lisa',
    }),
  ])
# ---
# name: test_apply_in_two_stages[second_migrations]
  list([
    dict({
      'applied_at': datetime,
      'created_at': datetime.datetime(2022, 6, 8, 16, 49, 51),
      'name': 'Rename username to user_id',
      'revision_id': 'mujt2zxouf9p',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime.datetime(2022, 6, 8, 16, 49, 52),
      'name': 'Add archived field',
      'revision_id': 'dnkkzcop90q0',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime.datetime(2022, 6, 8, 16, 49, 59),
      'name': 'Rename user_id to user',
      'revision_id': 'bq8byc2z1znq',
    }),
  ])
# ---
# name: test_apply_to_latest
  list([
    dict({
      'archived': False,
      'groups': list([
        'fred',
        'techs',
      ]),
      'user': 'fred',
    }),
    dict({
      'archived': False,
      'groups': list([
        'bob',
        'techs',
      ]),
      'user': 'bob',
    }),
    dict({
      'archived': False,
      'groups': list([
        'lisa',
      ]),
      'user': 'lisa',
    }),
  ])
# ---
# name: test_apply_to_latest.1
  list([
    dict({
      'applied_at': datetime,
      'created_at': datetime.datetime(2022, 6, 8, 16, 49, 51),
      'name': 'Rename username to user_id',
      'revision_id': 'mujt2zxouf9p',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime.datetime(2022, 6, 8, 16, 49, 52),
      'name': 'Add archived field',
      'revision_id': 'dnkkzcop90q0',
    }),
    dict({
      'applied_at': datetime,
      'created_at': datetime.datetime(2022, 6, 8, 16, 49, 59),
      'name': 'Rename user_id to user',
      'revision_id': 'bq8byc2z1znq',
    }),
  ])
# ---
